### PR TITLE
feat(repo): Initial content

### PR DESCRIPTION
**Pull request by micahhausler**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+RUN apk add -U ca-certificates
+
+ENV PKG=/go/src/github.com/skuid/keyman
+ADD . $PKG
+WORKDIR $PKG
+
+RUN go install -ldflags '-w'
+
+FROM alpine:latest
```
Comment (ethanfrogers) [inline Dockerfile]: `FROM alpine:3.6` for specificity.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+FROM golang:1.9-alpine as builder
+
+RUN apk add -U ca-certificates
```
Comment (ethanfrogers) [inline Dockerfile]: I feel like this should go into the actual image.

File: main.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		zap.L().Fatal("Error initializing logger", zap.Error(err))
+	}
+	zap.ReplaceGlobals(l)
+
+	// Viper binds
+	viper.BindPFlags(flag.CommandLine)
+	viper.SetEnvPrefix("keyman")
+	viper.AutomaticEnv()
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
```
Comment (ethanfrogers) [inline main.go:50]: We should consider making this a convenience method in out spec library.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	ca, err := sign.ReadPrivKey(viper.GetString("key"))
+	if err != nil {
+		zap.L().Fatal("Error reading key", zap.Error(err))
+	}
+
+	caServer := &server.Server{CA: ca}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/api/v1/sign", caServer.SignHandler)
```
Comment (ethanfrogers) [inline main.go]: I think `github.com/skuid/keyman/server` shoud have a method that returns a muxer for these handlers and just do `mux.Handle("/", caServer.Handlers())` or something so they they're contained within the pacakage.

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+type Server struct {
+	CA sign.Signer
+}
+
+func (s *Server) SignHandler(w http.ResponseWriter, r *http.Request) {
+
+	kr := &KeyRequest{}
+
+	err := json.NewDecoder(r.Body).Decode(kr)
```
Comment (ethanfrogers) [inline server/server.go]: `if err := json.NewDecoder(r.Body).Decode(kr); err != nil {...}`

File: server/server.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		http.Error(w, "Could not decode input", http.StatusBadRequest)
+		zap.L().Debug("Could not decode input", zap.Error(err))
+		return
+	}
+
+	// TODO change username, parameterize duration
+	signedCert, err := s.CA.Sign(
+		kr.Key,
+		"micahhausler",
```
Comment (ethanfrogers) [inline server/server.go]: see `// TODO`

File: server/server.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		kr.Principals,
+		(time.Duration(8) * time.Hour),
+	)
+	if err != nil {
+		http.Error(w, "Error signing request", http.StatusInternalServerError)
+		zap.L().Error("Error signing request", zap.Error(err))
+		return
+	}
+
+	err = json.NewEncoder(w).Encode(&KeyResponse{Certificate: signedCert})
```
Comment (ethanfrogers) [inline server/server.go]: `is err := json.NewEncoder(w).Encode(&KeyResponse{Certificate: signedCert}); err != nil {...}`

File: server/server.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	data := base64.StdEncoding.EncodeToString(pubKey.Marshal())
+	content := fmt.Sprintf("%s %s %s", "ssh-rsa", data, "ca@skuid.com")
+
+	response := &KeyResponse{
+		Certificate: content,
+	}
+
+	err = json.NewEncoder(w).Encode(response)
```
Comment (ethanfrogers) [inline server/server.go]: `if err := json.NewEncoder(w).Encode(response); err != nil {...}`

**Review by ethanfrogers — CHANGES_REQUESTED** [review]
Couple of stylistic changes.

File: Dockerfile (hunk 2 above)
Comment (micahhausler) [reply Dockerfile]: yep good call

File: Dockerfile (hunk 1 above)
Comment (micahhausler) [reply Dockerfile]: sure

File: main.go (hunk 3 above)
Comment (micahhausler) [reply main.go:50]: yea I'm not opposed to that

File: main.go (hunk 4 above)
Comment (micahhausler) [reply main.go]: cool

File: client/main.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	resp, err := client.Sign(context.Background(), request)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Certificate, nil
+}
+
+func main() {
+	flag.String("server", "localhost:3000", "The server to connect to")
+	flag.StringSlice("principals", []string{"core", "openvpnas"}, "The identities to request")
```
Comment (ethanfrogers) [inline client/main.go:33]: i don't know why, but i wish this was `principle`. When doing multiple, it would be `--principle core --principle openvpnas`.

Comment (ethanfrogers) [reply client/main.go:33]: your call though.

Comment (micahhausler) [reply client/main.go:33]: you can also do  `--principals root,core` so I'll just keep it

File: main.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+
+	ca, err := sign.ReadPrivKey(viper.GetString("key"))
+	if err != nil {
+		zap.L().Fatal("Error reading key", zap.Error(err))
+	}
+
+	caServer := &server.Server{
+		CA:        ca,
+		CaComment: "ca@skuid.com",
```
Comment (ethanfrogers) [inline main.go]: should this be configurable?

File: main.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+		Duration:  time.Duration(8) * time.Hour,
```
Comment (ethanfrogers) [inline main.go]: Same as above.

File: main.go (hunk 10 above)
Comment (micahhausler) [reply main.go]: yes

File: main.go (hunk 11 above)
Comment (micahhausler) [reply main.go]: sure

File: main.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		zap.L().Info("Metrics server is starting", zap.String("listen", metricsHostPort))
+		httpServer := &http.Server{Addr: metricsHostPort, Handler: internalMux}
+		lifecycle.ShutdownOnTerm(httpServer)
+		if err := httpServer.ListenAndServe(); err != http.ErrServerClosed {
+			zap.L().Fatal("Error listening", zap.Error(err))
+		}
+		zap.L().Info("Server gracefully stopped")
+	}()
+
+	grpcServer.Serve(lis)
```
Comment (ethanfrogers) [inline main.go]: No need to wrap with an `if err := ...`?

File: sign/ssh.go [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	err = cert.SignCert(c.Random, c.key)
+	if err != nil {
+		return nil, fmt.Errorf("Error signing cert: %q", err)
+	}
+
+	return []byte(fmt.Sprintf(
+		"%s %s %s",
+		"ssh-rsa-cert-v01@openssh.com",
```
Comment (ethanfrogers) [inline sign/ssh.go]: small nit: make this a constant?

Comment (micahhausler) [reply sign/ssh.go]: ah yea, this should actually be conditional: https://godoc.org/golang.org/x/crypto/ssh#pkg-constants

Comment (ethanfrogers) [reply sign/ssh.go]: 👍

**Review by ethanfrogers — APPROVED** [review]
LGTM. Just a couple of small nits.

**Pull request merged**